Sources/TurboNavigator.swift

```diff
@@ -1,235 +1,62 @@
-                modalNavigationController: UINavigationController = UINavigationController())
-    {
-        self.session = Session(webView: TurboConfig.shared.makeWebView())
-        self.modalSession = Session(webView: TurboConfig.shared.makeWebView())
-        self.delegate = delegate
-        self.navigationController = navigationController
-        self.modalNavigationController = modalNavigationController
+public class TurboNavigator: TurboNavigationHierarchyControllerDelegate {
+    public unowned var delegate: TurboNavigatorDelegate
 
-        session.delegate = self
-        modalSession.delegate = self
-        session.pathConfiguration = pathConfiguration
-        modalSession.pathConfiguration = pathConfiguration
+    public var rootViewController: UINavigationController { hierarchyController.navigationController }
+
+    public var webkitUIDelegate: TurboWKUIController? {
+        didSet {
+            session.webView.uiDelegate = webkitUIDelegate
+            modalSession.webView.uiDelegate = webkitUIDelegate
+        }
     }
 
-    /// Provide `Turbo.Session` instances with preconfigured path configurations and delegates.
-    /// Note that TurboNavigationDelegate.controller(_:forProposal:) will no longer be called.
-    /// - Parameters:
-    ///   - preconfiguredMainSession: a session whose delegate is not `TurboNavigator`
-    ///   - preconfiguredModalSession: a session whose delegate is not `TurboNavigator`
-    ///   - delegate: handle non-routing behavior, like custom error handling
-    ///   - navigationController: optional: override the main navigation stack
-    ///   - modalNavigationController: optional: override the modal navigation stack
-    public init(preconfiguredMainSession: Turbo.Session,
-                preconfiguredModalSession: Turbo.Session,
-                delegate: TurboNavigationDelegate,
-                navigationController: UINavigationController = UINavigationController(),
-                modalNavigationController: UINavigationController = UINavigationController())
-    {
-        self.session = preconfiguredMainSession
-        self.modalSession = preconfiguredModalSession
-        self.navigationController = navigationController
-        self.modalNavigationController = modalNavigationController
+    public init(session: Session, modalSession: Session, delegate: TurboNavigatorDelegate? = nil) {
+        self.session = session
+        self.modalSession = modalSession
+        
+        self.delegate = delegate ?? navigatorDelegate
 
-        self.delegate = delegate
-    }
+        self.session.delegate = self
+        self.modalSession.delegate = self
 
-    public var rootViewController: UIViewController { navigationController }
-    public let navigationController: UINavigationController
-    public let modalNavigationController: UINavigationController
+        defer { self.webkitUIDelegate = TurboWKUIController(delegate: self) }
```

Comment:
`defer` is needed to trigger the `didSet` callback above. I don't like this but can't think of another way without duplicating the logic in that callback.

Reply:
I don't mind duplicating the logic. It really should just be about setting the delegate to both sessions and that's what we're doing here already. I don't feel strongly about it, though.